### ISSUE TITLE
Proposal: opaque keyword

### ISSUE THREAD

**Issue by tadeokondrak** [ziglang/zig#5574]
Turn
```zig
const CType = @Type(.Opaque);
```
into
```zig
const CType = opaque{};
```
and allow those opaque types to have declarations like constants/functions just like structs, with the difference that you can only use it as a pointer.

This is useful for wrapping C APIs, from my attempts you can't really fake this with a struct and keep using pointers to the type, so you have to use a `struct { ptr: *c_type }` which hides the pointer-ness of the type and removes the ability to use const.

**Comment by deingithub** [comment]
Instead of introducing a new keyword, but still use this IMO very worthwhile proposal, one could change the tag of `std.builtin.TypeInfo.Opaque` to be a `type` (a struct in our case) instead of `void`, and use it like this:
```zig
const CType = @Type(.Opaque = struct {
    pub const adjective = "questionable";
    pub fn eels() void {}
});
```
The new opaque type would use all declarations from the passed-in struct. Giving this struct fields would be a compile error as it is not meant to be ever instantiated.

**Comment by tadeokondrak** [comment]
> Instead of introducing a new keyword, but still use this IMO very worthwhile proposal, one could change the tag of std.builtin.TypeInfo.Opaque to be a type (a struct in our case) instead of void

How would you access `@This()`?

An alternative to using `opaque` as a keyword would be to use `opaquetype`.

**Comment by deingithub** [comment]
> How would you access `@This()`?
Admittedly, I hadn't thought about that. Since we can't use the name of the thing we're declaring in this moment, the approach might not work out without the introduction of another internal function like `@OpaqueThis()` or a change to the semantics of `@This()`, both of which aren't really much cleaner than adding a new keyword anymore — I'll leave the post up in case someone has an idea how to handle that better.

**Comment by MasterQ32** [comment]
I like the idea of having method syntax on opaque C types. This allows simpler wrappers to be written, but introducing a new keyword for this is … dissatisfying

Half-thumbs-up from my side.

Brainfart: Right now, `struct`s are `0`-sized types whereas opaque types are `null`-sized types (meaning you cannot take `sizeOf`). Not sure we should change this, but having an option to define pointer-only types is quite helpful

**Comment by andrewrk** [comment]
Implemented in #6421 by @tadeokondrak
